wallet/reservation.c

```diff
@@ -164,3 +165,115 @@ static const struct json_command unreserveinputs_command = {
 	false
 };
 AUTODATA(json_command, &unreserveinputs_command);
+
+
+static struct command_result *json_fundpsbt(struct command *cmd,
+					      const char *buffer,
+					      const jsmntok_t *obj UNNEEDED,
+					      const jsmntok_t *params)
+{
+	struct json_stream *response;
+	const struct utxo **utxos;
+	u32 *feerate_per_kw;
+	u32 *minconf;
+	struct amount_sat *amount, input, needed, excess;
+	bool all;
+	u32 locktime, maxheight;
+	struct bitcoin_tx *tx;
+
+	if (!param(cmd, buffer, params,
+		   p_req("satoshi", param_sat_or_all, &amount),
+		   p_req("feerate", param_feerate_val, &feerate_per_kw),
+		   p_opt_def("minconf", param_number, &minconf, 1),
+		   NULL))
+		return command_param_failed();
+
+	all = amount_sat_eq(*amount, AMOUNT_SAT(-1ULL));
+	maxheight = minconf_to_maxheight(*minconf, cmd->ld);
+
+	/* We keep adding until we meet their output requirements. */
+	input = AMOUNT_SAT(0);
+	utxos = tal_arr(cmd, const struct utxo *, 0);
+	while (amount_sat_sub(&needed, *amount, input)) {
+		struct utxo *utxo;
+
+		utxo = wallet_find_utxo(utxos, cmd->ld->wallet,
+					cmd->ld->topology->tip->height,
+					&needed,
+					*feerate_per_kw,
+					maxheight,
+					utxos);
+		if (utxo) {
+			struct amount_sat fee;
+			tal_arr_expand(&utxos, utxo);
+
+			/* It supplies more input. */
+			if (!amount_sat_add(&input, input, utxo->amount))
+				return command_fail(cmd, LIGHTNINGD,
+						    "impossible UTXO value");
+
+			/* But increase amount needed, to pay for new input */
+			fee = amount_tx_fee(*feerate_per_kw,
+					    utxo_spend_weight(utxo));
+			if (!amount_sat_add(amount, *amount, fee))
+				/* Either they specified "all", or we
+				 * will fail anyway. */
+				*amount = AMOUNT_SAT(-1ULL);
+			continue;
+		}
+
+		/* If they said "all", we expect to run out of utxos. */
+		if (all)
+			break;
+
+		return command_fail(cmd, FUND_CANNOT_AFFORD,
+				    "Could not afford %s using all %zu available UTXOs: %s short",
+				    type_to_string(tmpctx,
+						   struct amount_sat,
+						   amount),
+				    tal_count(utxos),
+				    type_to_string(tmpctx,
+						   struct amount_sat,
+						   &needed));
+	}
+
+	/* Setting the locktime to the next block to be mined has multiple
+	 * benefits:
+	 * - anti fee-snipping (even if not yet likely)
+	 * - less distinguishable transactions (with this we create
+	 *   general-purpose transactions which looks like bitcoind:
+	 *   native segwit, nlocktime set to tip, and sequence set to
+	 *   0xFFFFFFFE by default. Other wallets are likely to implement
+	 *   this too).
+	 */
+	locktime = cmd->ld->topology->tip->height;
+
+	/* Eventually fuzz it too. */
+	if (locktime > 100 && pseudorand(10) == 0)
+		locktime -= pseudorand(100);
+
+	/* FIXME: tx_spending_utxos does more than we need, but there
+	 * are other users right now. */
+	tx = tx_spending_utxos(cmd, chainparams, utxos,
+			       cmd->ld->wallet->bip32_base,
+			       false, 0, locktime,
+			       BITCOIN_TX_RBF_SEQUENCE);
```

Comment:
Why RBF-flagged?  Am I able to *actually* RBF now??  How does RBF work?

Reply:
Yes, we can RBF once everyone uses this...

Reply:
Okay, let me see if my understanding is correct.

To implement RBF, I would have to go all the way up to `signpsbt`, but I would have to *independently* extract the transaction and broadcast it, such as by the `bcli` `sendrawtransaction` method.  Then if I decide to RBF, I would mutate the PSBT directly, then call `signpsbt` again with the RBFed PSBT, then `sendrawtransaction` again.

(we cannot call `sendpsbt` more than once, since it marks consumed coins as spent, with an assertion that the consumed coin must currently be reserved)

Then, if any of the transactions get confirmed onchain, this automatically marks the consumed inputs as spent.  Does that happen?

Looking at `lightningd/chaintopology.c` it  seems we have the rough framework to do so, but while it calls into `wallet_outpoint_spend`, that function does not update the `status` of UTXOs.  It also calls into `wallet_extract_owned_outputs`, which will get any unknown outputs, but not what a transaction spends.

`wallet_update_update_output_status` is otherwise only called from within `wallet/wallet.c`, or from `dev-rescan-outputs`.

I don't see any changes in `wallet/wallet.c` that would allow the above flow to be used.

Reply:
And then there is the issue where one version of the transaction is seen by one miner, the other version is seen by the other miner, then we see the block created by the first, but ultimately the second miner gets a second block, so now we have to rewind our chaintopology and destroy UTXOs in that chaintopology and recreate it as we go to the winning branch.  Do we have code that handles that?  Because if not, we are not RBF safe.